src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForBuilderDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,267 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.LanguageService;
+using Microsoft.CodeAnalysis.CSharp.Shared.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.UseCollectionInitializer;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.UseCollectionInitializer;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseCollectionExpression;
+
+internal readonly record struct CollectionBuilderMatch(
+    Location DiagnosticLocation,
+    LocalDeclarationStatementSyntax LocalDeclarationStatement,
+    InvocationExpressionSyntax CreationExpression,
+    ImmutableArray<Match<StatementSyntax>> Matches);
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+internal sealed partial class CSharpUseCollectionExpressionForBuilderDiagnosticAnalyzer
+    : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+{
```

Comment:
all the code from here to AnalyzeInvocation is just the boilerplate we have for all our analyzers here.

Reply:
note: they're so common, i may just merge them into one base class at some point.